dist/tools: add firmware metadata generator

kYc0o opened the pull request:
Since #6450 became a bit complicated to manage, I'll start to split the work in several PRs, with the aim to ease review and have a better picture of the approach.

This PR adds a firmware metadata generator, which will produce a binary file representing a structure of this kind:

```c
typedef struct FW_metadata {
    uint8_t hash[SHA256_DIGEST_LENGTH]; /**< SHA256 Hash of firmware image */
    uint8_t shash[SIGN_LEN];            /**< Signed SHA256 */
    uint16_t version;                   /**< Integer representing firmware version */
    uint32_t size;                      /**< Size of firmware image */
    uint32_t appid;                     /**< Integer representing the application ID */
} FW_metadata_t;
```

as a simplified way to characterise the contents of a firmware, while adding some security features as the sha256 hash for integrity check, as well as a field for a signed hash, which will come in a following PR.

This approach will use the current RIOT implementation of sha256 to allow portability.

To test the firmware generator, you can run the following command for any RIOT example or application:

`make firmware-metadata VERSION=0x1 APPID=0xabcd1234`

using `VERSION=0x1` and `APPID=0xabcd1234` as examples.

The ultimate goal is to put this metadata at the beginning of any firmware, to use it e.g. for a firmware update. This feature will come in an upcoming PR, stay tuned.

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,163 @@
+
+#define FW_METADATA_BIN "firmware-metadata.bin"
+
+typedef struct FW_metadata {
```
This doesn't seem to follow coding conventions, why not use the explicit name: `firmware_metadata` ? and `firmware_metadata_t` below ?

aabadie commented on Makefile.include:
```diff
@@ -303,9 +306,18 @@ else
 endif # BUILD_IN_DOCKER
 
+firmware-metadata: all generate-metadata
+	$(Q)$(FW_METADATA)/bin/./generate-metadata $(BINFILE) $(VERSION) $(APPID) $(METADATA_FILE)
```
do you really need the `./` in `$(Q)$(FW_METADATA)/bin/./generate-metadata` ?

aabadie commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+all: bin bin/generate-metadata
+
+bin:
+	mkdir bin
```
maybe use `-p` to avoid messages like `mkdir: cannot create directory ‘bin’: File exists`

aabadie commented on dist/tools/firmware_metadata/README.md:
````diff
@@ -0,0 +1,35 @@
+follows:
+
+```c
+typedef struct FW_metadata {
````
same comment as below about structure name

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2016, Mark Solters <msolters@gmail.com>.
+ *               2016, Francisco Acosta <francisco.acosta@inria.fr>
```
I think you should simply put Inria in the copyright

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,163 @@
+    fclose(firmware_bin);
+
+    /*
+     * TODO Sign hash
```
Do you plan to address this ?

kYc0o replied:
Yes, it will come in a following up PR.

aabadie replied:
So I guess it can be removed from this one.

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    /* Generate FW image metadata */
+
```
this empty line is not needed

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,163 @@
+
+    (void)argc;
+
+    if (!argv[1]) {
```
Can you just check `argc` value and display the usage in case it's < 4 ?

Kijewski commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+
+all: bin bin/generate-metadata
+
+bin:
```
add a slash `bin/`

Kijewski commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+
+CFLAGS += -g -O3 -Wall -Wextra -pedantic -std=c99
+
+all: bin bin/generate-metadata
```
remove `bin`

Kijewski commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+bin:
+	mkdir bin
+
+bin/generate-metadata: $(METADATA_HDR) $(METADATA_SRC)
```
append `| bin/`

kYc0o commented:
@aabadie @Kijewski Comments addressed.

Kijewski commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+bin/:
+	mkdir -p bin
+
+bin/generate-metadata: bin/ $(METADATA_HDR) $(METADATA_SRC)
```
`bin/` should be a pre-requirement instead of a requirement. A pre-requirement only needs to exist, but a newer timestamp does not force a rebuild. Creating `bin/generate-metadata` will touch `bin/`, which will tell `make` to recreate `bin/generate-metadata` in its next invocation. Pre-requirements are placed after the requirements, separated by a pipe.

kYc0o replied:
Oh! Ok I understand now. However, don't we actually want `bin` as a requirement? What's really the difference? Anyways,`bin` can only exists if a build was triggered, and if I understand correctly the contents cannot be changed by other means.

Kijewski replied:
When a file is changed, then the modification timestamp of its parent folder gets updated. A newer timestamp of a requirement $X tells Make to rebuild $X. So updating $X will make $X look outdated to Make. Every time you run Make it will update $X, which wastes a tiny amount of wall clock time.

As a rule of thumb directories should always be prerequirements, not requirements, unless you are doing something really odd. :)

kYc0o replied:
OK, thanks for the explanation, I'll modify this file accordingly.

aabadie commented on dist/tools/firmware_metadata/Makefile:
```diff
@@ -0,0 +1,19 @@
+RIOTBASE := ../../..
```
I think this line should be:
```
RIOTBASE ?= $(CURDIR)/../../..
```

kYc0o replied:
I think since dist/tools/ is a path that will always exists in any clone of RIOT, we don't really need to define `$(CURDIR)`, and so the relative path will always be the same.

aabadie replied:
Ok you are right.

kYc0o commented:
@Kijewski addressed.

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,35 @@
+# Metadata generator for firmware verification
+This program will generate a binary file containing a metadata structure as
```
Minor: use present instead of future.
`This program generates a binary file...`

aabadie commented on dist/tools/firmware_metadata/README.md:
````diff
@@ -0,0 +1,35 @@
+} firmware_metadata_t;
+```
+
+This structure will be filled with the data obtained from the firmware.
````
Same here: `This structure is filled with...`

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,35 @@
+
+_\<output-path\>_\: The path for fimrware_metadata.bin
+
+The results will be printed if the operation is successful, and a binary
```
Same, with `if` the sense is conditional (something like might be). Also use present:
`If the operation succeeds, the results is printed out and a binary called "firmware-metadata.bin" is created (written?)`

aabadie replied:
Looking at `generate-metadata.c` this is not totally true: the user can set a custom output filename.

aabadie commented:
@kYc0o, what's the status here ?

kYc0o commented:
Just addressed your comments.

kYc0o commented:
@Kijewski all ok for you?

aabadie commented:
@kYc0o, Murdock is reporting errors, can you have a look ?

kYc0o commented:
It's again related to the pic32 boards... @kaspar030 there is something wrong in the toolchain of Murdock2?

BTW it needs rebase.

kYc0o commented:
> BTW it needs rebase.

I meant squash.

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+
+_\<appid\>_\: ID for the application in 32-bit HEX
+
+_\<output-path\>_\: The path for fimrware_metadata.bin
```
firmware_metadata.bin => fi**r**mware_metadata.bin

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+
+If the operation succeeds, the results are printed out and a binary called
+"firmware-metadata.bin" is written, if no _output-path_ option is specified,
+in which case the file is written to the given path with the given name.
```
`in which case` should be removed

aabadie commented:
Just tested it and had problems:
```
bin/generate-metadata ~/samr21-default.elf 1.0 1234 ~/.
```
generate a segfault. Here is what gdb says:
```
Starting program: /home/aabadie/softs/src/riot/RIOT/dist/tools/firmware_metadata/bin/generate-metadata /home/aabadie/samr21-default.elf 1.0 1234 /home/aabadie/.
Firmware bytes read: 1855452
Firmware Size: 1855452
Firmware Version: 0x1
Firmware APPID: 0x1234
Firmware HASH: ce 9c f5 56 84 44 2f 36 97 50 1f 6b 34 68 56 3f 4c 5d dc a9 a9 48 71 73 59 08 b8 1a ac 65 80 b3 
Firmware signed HASH: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
Metadata size: 76

Program received signal SIGSEGV, Segmentation fault.
__GI__IO_fwrite (buf=0x7fffffffd1d0, size=76, count=1, fp=0x0) at iofwrite.c:37
37	iofwrite.c: No such file or directory.
```

Leaving the output path empty works fine.

miri64 commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+
+Where:
+
+_\<firmware.bin\>\:_ The firmware in binary format
```
Maybe use `<firmware.bin>` instead so it is better readable in raw text?

kYc0o replied:
I don't get exactly what you mean...

aabadie replied:
I think @kaspar030 means "use \` \` around \<firmware.bin\>". +1 on my side

kYc0o commented:
@aabadie addressed comments.

Your problem is that you didn't give a file name for the binary with the metadata. I fixed the README.md to be explicit about this.

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+_\<output-path/filename.bin\>_\: The path and name of the metadata binary file
+
+If the operation succeeds, the results are printed out and a binary called
+"firmware-metadata.bin" is written, if no _output-path_ option is specified.
```
This is confusing as first sight.

kYc0o replied:
What's the confusion?

aabadie commented on dist/tools/firmware_metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+# Metadata generator for firmware verification
+This program generates a binary file containing a metadata structure as
```
Maybe add an empty line after title.

Quick question: does the file only contain the metadata or the metadata + the initial firmware ? If it's the second case, it should be more explicit. Otherwise, one can easily understand that the generated binary file only contain the struct.

kYc0o replied:
> Maybe add an empty line after title.

What this will change?

> uick question: does the file only contain the metadata or the metadata + the initial firmware ? If it's the second case, it should be more explicit. Otherwise, one can easily understand that the generated binary file only contain the struct.

It is the first case.

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,153 @@
+    char firmware_metadata_path[128];
+
+    if (argc < 4) {
+        puts("Usage: generate-metadata <BINFILE> <VERSION> <APPID> [output path]");
```
s/output path/output filename/

kYc0o replied:
I still think output path is ok...

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,153 @@
+
+    sha256_init(&firmware_sha256);
+
+    while((bytes_read = fread(firmware_buffer, 1, sizeof(firmware_buffer), firmware_bin))) {
```
space after while missing

aabadie commented on dist/tools/firmware_metadata/generate-metadata.c:
```diff
@@ -0,0 +1,153 @@
+/* Output metadata .bin file */
+FILE *metadata_bin;
+
+uint32_t firmware_size = 0;
```
could be declared static

kYc0o replied:
Does it matter for a native C program? In that case all the variables should be declared static in this program.

kYc0o commented on dist/tools/firmware_metadata/README.md:
````diff
@@ -0,0 +1,36 @@
+To use, you should call `generate-metadata` with the following arguments:
+
+```console
+./generate-metadata <firmware.bin> <version> <appid> <output-path>
````
I think `<output-filename>` would lead to confusion, since we need the full path _plus_ the file name.

kYc0o commented:
Actually, I don't think that _beautify_ a lot the README would be of any use. This tool is not really meant to be used by a final user, but called automatically by the build system when building firmware updates like in the subsequent PRs.

kYc0o commented:
> Let's not merge this until we have a set of PR's that actually enable OTA.

What do you mean with that? I'm about to push the last PR which enables the "updates fetch" using CoAP (though the transport method is TFTP). That would be the last PR regarding this topic.

kYc0o commented:
#7167 adds the firmware downloading for OTA updates. I think with this PR the firmware update functionality is complete.

kYc0o commented:
Closing in favour of #7457.
